### ISSUE TITLE
focus on listbox move jumps

### ISSUE THREAD

**Issue by kkrolczyk** [urwid/urwid#357]
##### Description:
When using up/down keys to move / select items in listbox, focus jumps to first invisible row, after screen. Focus does not move if listbox is displayed completely. 

##### Affected versions (if applicable)
- [x] `master` branch (bb5498690bb5)
- [ ] Latest stable version from `pypi`
- [ ] Other (specify source)

##### Steps to reproduce (if applicable)

    import urwid
    import sys

    palette = {
        ("n",   "light red",    "default",      "default"),
        ("f",   "light green",  "white",        "default"),
    }
    def exit_on_(key):
        if key in ("q", "Q", "esc"):
            raise urwid.ExitMainLoop()

    class SomeListBox(urwid.ListBox):
        def __init__(self, cnt):
            x = [urwid.AttrMap(urwid.Columns([
                 urwid.Text(str(j*i)) for j in (2, 3)]), "n", "f")
                 for i in range(cnt)]

            body = urwid.SimpleFocusListWalker(x)
            super(SomeListBox, self).__init__(body)

        def keypress(self, size, key):
            key = super(SomeListBox, self).keypress(size, key)
            if key in ("up", "down"):
                # jumps and focuses off screen row.
                return key
            return key

    try:
        cnt = int(sys.argv[1])
    except:
        cnt = 100

    frame = urwid.Frame(SomeListBox(cnt))
    loop = urwid.MainLoop(frame, palette, unhandled_input=exit_on_)
    loop.run()

##### Expected/actual outcome
focus moves and iterates row, row+1 regardless of list visibility

Problem possibly originates in listbox.py line 1013

        # look for selectable widget above
        pos = focus_pos
        widget = None
        for widget, pos, rows in fill_above:
            row_offset -= rows
            if rows and widget.selectable():
                # this one will do
                self.change_focus((maxcol,maxrow), pos,
                    row_offset, 'below')
                return
 
The code does not seem to be executed for me, with example above widget.selectable() is False. Code above somewhat works, when for used "if rows and (1 or widget.selectable())".
But even then, focus jumps 2*lines, so: first focus moves +1  line, next +2 lines (from 2 to 4), next +4 lines (from 4 to 8) ect.

**Comment by tonycpsu** [comment]
TL:DR: I think this is expected behavior.

The default implementation for up and down keystrokes in urwid widgets is to move to the next selectable widget.  If there are no selectable widgets above or below (depending on key),  and if you're not overriding the listbox's keypress method, then the `keypress_up` and `keypress_down` methods in `listbox.py` will, instead of doing nothing, scroll the first or last widget.

Without knowing what went into that decision, my guess is that moving to the top/bottom then allows subsequent up/down presses to be passed to container widgets above/below the listbox.

To get the behavior you seem to want, you can just make the widget you're using for listbox entries selectable:

```python
import urwid
import sys

palette = {
    ("n",   "light red",    "default",      "default"),
    ("f",   "light green",  "white",        "default"),
}
def exit_on_(key):
    if key in ("q", "Q", "esc"):
        raise urwid.ExitMainLoop()

class Row(urwid.WidgetWrap):

    def __init__(self, i):
        self.columns = urwid.Columns([
            urwid.Text(str(j*i)) for j in (2, 3)
        ])
        self.attr = urwid.AttrMap(self.columns, "n", "f")
        super().__init__(self.attr)

    def selectable(self):
        return True

class SomeListBox(urwid.ListBox):
    def __init__(self, cnt):
        x = [Row(i) for i in range(cnt)]
        body = urwid.SimpleFocusListWalker(x)
        super(SomeListBox, self).__init__(body)

try:
    cnt = int(sys.argv[1])
except:
    cnt = 100

frame = urwid.Frame(SomeListBox(cnt))
loop = urwid.MainLoop(frame, palette, unhandled_input=exit_on_)
loop.run()
```

Alternately, you could just handle up/down explicitly in your `ListBox` subclass, setting `focus_position`.  But I think the above is easier.

**Comment by kkrolczyk** [comment]
Thanks @tonycpsu I fully agree, with Your example it works OK, but then I unfortunately overdone my minimal-viable-example - real usage in my app is quite a bit more convoluted. I had somewhere less contrived idea showing what I wanted to achieve, but for now I am quite swamped with work and had no time for that. I'll revisit this though.
